### PR TITLE
[SPARK-14125] [SQL] Native DDL Support: Alter View

### PR DESCRIPTION
**Pull request by gatorsmile**
#### What changes were proposed in this pull request?

This PR is to provide a native DDL support for the following three Alter View commands:

Based on the Hive DDL document:
https://cwiki.apache.org/confluence/display/Hive/LanguageManual+DDL
##### 1. ALTER VIEW RENAME

**Syntax:**

``` SQL
ALTER VIEW view_name RENAME TO new_view_name
```
- to change the name of a view to a different name
- not allowed to rename a view's name by ALTER TABLE
##### 2. ALTER VIEW SET TBLPROPERTIES

**Syntax:**

``` SQL
ALTER VIEW view_name SET TBLPROPERTIES ('comment' = new_comment);
```
- to add metadata to a view
- not allowed to set views' properties by ALTER TABLE
- ignore it if trying to set a view's existing property key when the value is the same
- overwrite the value if trying to set a view's existing key to a different value
##### 3. ALTER VIEW UNSET TBLPROPERTIES

**Syntax:**

``` SQL
ALTER VIEW view_name UNSET TBLPROPERTIES [IF EXISTS] ('comment', 'key')
```
- to remove metadata from a view
- not allowed to unset views' properties by ALTER TABLE
- issue an exception if trying to unset a view's non-existent key
#### How was this patch tested?

Added test cases to verify if it works properly.

### CODE REVIEW

**Comment by gatorsmile** [comment]
cc @yhuai @andrewor14

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 1]
```diff
@@ -232,16 +232,17 @@ case class DropTable(
...
   extends RunnableCommand {
 
   override def run(sqlContext: SQLContext): Seq[Row] = {
     val catalog = sqlContext.sessionState.catalog
+    DDLUtils.verifyAlterTableOrAlterView(catalog, oldName, isView)
     catalog.invalidateTable(oldName)
     catalog.renameTable(oldName, newName)
     Seq.empty[Row]
   }
-
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: nit: I left this line intentionally

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 2]
```diff
@@ -533,5 +538,23 @@ private object DDLUtils {
   def isDatasourceTable(table: CatalogTable): Boolean = {
     isDatasourceTable(table.properties)
   }
+
+  // If the command ALTER VIEW is to alter a table or ALTER TABLE is to alter a view,
+  // issue an exception.
+  @throws[AnalysisException]
+  def verifyAlterTableOrAlterView(
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I would just call this `verifyAlterTableType` and make this private

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 3]
```diff
@@ -533,5 +538,23 @@ private object DDLUtils {
   def isDatasourceTable(table: CatalogTable): Boolean = {
     isDatasourceTable(table.properties)
   }
+
+  // If the command ALTER VIEW is to alter a table or ALTER TABLE is to alter a view,
+  // issue an exception.
+  @throws[AnalysisException]
```
Comment (andrewor14) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: can you turn these into a java doc?

**Comment by andrewor14** [comment]
Looks great! Only minor comments.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 3 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Sure, will do it. Thanks!

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 2 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Will rename it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala (hunk 1 above)
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Ok, I see.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 4]
```diff
@@ -131,6 +131,124 @@ class HiveDDLSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+        val catalog = hiveContext.sessionState.catalog
+        sql(s"CREATE VIEW $oldViewName AS SELECT * FROM $tabName")
+
+        assert(catalog.tableExists(TableIdentifier(oldViewName)))
+        assert(!catalog.tableExists(TableIdentifier(newViewName)))
+        sql(s"ALTER VIEW $oldViewName RENAME TO $newViewName")
+        assert(!catalog.tableExists(TableIdentifier(oldViewName)))
+        assert(catalog.tableExists(TableIdentifier(newViewName)))
+
+        sql(s"DROP VIEW $newViewName")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: `withView` will help us drop the view at last.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 5]
```diff
@@ -131,6 +131,124 @@ class HiveDDLSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+        assert(catalog.getTableMetadata(TableIdentifier(viewName))
+          .properties.filter(_._1 != "transient_lastDdlTime") == Map())
+
+        val message = intercept[AnalysisException] {
+          sql(s"ALTER VIEW $viewName UNSET TBLPROPERTIES ('p')")
+        }.getMessage
+        assert(message.contains(
+          "attempted to unset non-existent property 'p' in table '`view1`'"))
+
+        sql(s"DROP VIEW $viewName")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: here too

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala [hunk 6]
```diff
@@ -131,6 +131,124 @@ class HiveDDLSuite extends QueryTest with SQLTestUtils with TestHiveSingleton {
...
+        message = intercept[AnalysisException] {
+          sql(s"ALTER TABLE $oldViewName UNSET TBLPROPERTIES ('p')")
+        }.getMessage
+        assert(message.contains(
+          "Cannot alter a view with ALTER TABLE. Please use ALTER VIEW instead"))
+
+        assert(catalog.tableExists(TableIdentifier(tabName)))
+        assert(catalog.tableExists(TableIdentifier(oldViewName)))
+
+        sql(s"DROP VIEW $newViewName")
```
Comment (cloud-fan) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: here too

**Comment by cloud-fan** [comment]
LGTM except some minor comments

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala (hunk 4 above)
Comment (gatorsmile) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveDDLSuite.scala]: Sure, remove it now. Thanks!

**Comment by yhuai** [comment]
Thanks. Merging to master.

**Pull request closed without merging**